### PR TITLE
Allow non-violent pawns to use non-lethal weapons

### PR DESCRIPTION
**Pull request by LockdownX7**
## Additions
- Any weapon containing `CE_Nonlethal_Melee` or `CE_Nonlethal_Ranged` in its weaponTags can now be used by non-violent pawns.

## Notes
- I had to set the aiming stats to neverDisabled to prevent error messages, as non-violent pawns need these to use grenades and other ranged weapons that might be set to non-lethal later on. As far as I'm aware, this should not have any negative effects, it just means that non-violent pawns now have aiming accuracy displayed in their info window.

### CODE REVIEW

**Review by N7Huntsman — COMMENTED** [review]
It does allow them to equip non-lethal weapons, but it also lets them punch things, and I'm not sure what we think of that.

File: Source/CombatExtended/Harmony/Harmony_FloatMenuMakerMap.cs [hunk 1]
```diff
@@ -417,4 +417,36 @@ static bool ForceWearInventoryCheck(Pawn pawn, Apparel apparel, List<FloatMenuOp
...
+                    patchApplied = true;
+                    break;
+                }
+            }
+            if (!patchApplied)
+            {
+                Log.Error($"{MethodBase.GetCurrentMethod().DeclaringType} did not find opcode to apply transpiler patch.");
+            }
+            return instructionsArray;
+        }
```
Comment (MadaraUchiha) [inline Source/CombatExtended/Harmony/Harmony_FloatMenuMakerMap.cs:445]: This whole thing can be simplified by using Harmony's `MethodReplacer` extension method. https://github.com/pardeike/Harmony/blob/master/Harmony/Public/Transpilers.cs#L19

```suggestion
    class Harmony_FloatMenuMakerMap_AddHumanlikeOrders_Patch
    {
        [HarmonyTranspiler]
        static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions)
        {
            return instructions.MethodReplacer(AccessTools.PropertyGetter(typeof(ThingDef), nameof(ThingDef.IsWeapon)), AccessTools.Method(typeof(Harmony_FloatMenuMakerMap_AddHumanlikeOrders_Patch), nameof(IsLethalWeapon)));
        }
```

Note that I have not remotely tested this. It Should™ Work™ but if you do go that route test it thoroughly.

File: Source/CombatExtended/Harmony/Harmony_Pawn.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        public static IEnumerable<MethodBase> TargetMethods()
+        {
+            yield return AccessTools.Method(typeof(Pawn), nameof(Pawn.TryStartAttack));
+            yield return AccessTools.Method(typeof(VerbTracker), "CreateVerbTargetCommand");
+            yield return AccessTools.Method(typeof(FloatMenuUtility), nameof(FloatMenuUtility.GetRangedAttackAction));
+            yield return AccessTools.Method(typeof(FloatMenuUtility), nameof(FloatMenuUtility.GetMeleeAttackAction));
+        }
+
+        [HarmonyTranspiler]
+        static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions)
```
Comment (MadaraUchiha) [inline Source/CombatExtended/Harmony/Harmony_Pawn.cs:27]: Same comment as last one

**Review by MadaraUchiha — APPROVED** [review]
I don't want to block this if it's considered urgent, and the current form will work properly, but do consider the comments :)

**Comment by N7Huntsman** [comment]
> I don't want to block this if it's considered urgent, and the current form will work properly, but do consider the comments :)

I don't consider this change to be particularly urgent or even strictly necessary, depending on how much effort it would take to implement in a satisfactory way. While the idea of non-violent colonists using non-lethal equipment is enticing, especially for stuff like tools or firefighting equipment, I think the utter inability to fight is an interesting gameplay aspect/challenge, and I don't know if I the equipment is worthwhile enough to come the cost of throwing away that aspect in exchange for turning them into incredibly shitty melee fighters.

**Comment by P4V3L-3D0TS** [comment]
what about 40mm beanbag ammo?

**Comment by LockdownX7** [comment]
The non-lethal tag has been split into `CE_Nonlethal_Melee` and `CE_Nonlethal_Ranged` tags, allowing more control over how non-violent pawns use them.

**Comment by Reviire** [comment]
> It does allow them to equip non-lethal weapons, but it also lets them punch things, and I'm not sure what we think of that.

While technically not correct behaviour it's such a minor thing that will have very little gameplay impact outside of abusing certain combinations of gear to have unarmed powerhouses. At which point... the player is knowingly exploiting something. I don't think there'd be any major consequences to this change that the player.

**Comment by N7Huntsman** [comment]
Closed, but branch left open in case we want to implement this capability later on.

**Pull request closed without merging**